File: go/running_stat/running_stat.go
package main

import "math"

type runningStat struct {
	count   int
	oldMean float64
	newMean float64
	oldSum  float64
	newSum  float64
}

func (rs *runningStat) clear() {
	rs.count = 0
	rs.oldMean = 0
	rs.newMean = 0
	rs.oldSum = 0
	rs.newSum = 0
}

func (rs *runningStat) mean() float64 {
	if rs.count > 0 {
		return rs.newMean
	} else {
		return 0
	}
}

func (rs *runningStat) variance() float64 {
	if rs.count > 1 {
		return rs.newSum / float64(rs.count-1)
	} else {
		return 0
	}
}

func (rs *runningStat) standardDeviation() float64 {
	return math.Sqrt(rs.variance())
}

func (rs *runningStat) push(value float64) {
	rs.count++

	if rs.count == 1 {
		rs.oldMean = value
		rs.newMean = value
		rs.oldSum = 0.0
	} else {
		rs.newMean = rs.oldMean + (value-rs.oldMean)/float64(rs.count)
		rs.newSum = rs.oldSum + (value-rs.oldMean)*(value-rs.newMean)

		// set up for next iteration
		rs.oldMean = rs.newMean
		rs.oldSum = rs.newSum
	}
}

File: dotnet/runningStat/net7/RunningStat.cs
namespace Stat;

public class RunningStat
{
    int count;
    double old_mean;
    double new_mean;
    double old_sum;
    double new_sum;

    public RunningStat()
    {
        this.Clear();
    }

    public Task Clear()
    {
        this.count = 0;
        this.old_mean = 0;
        this.new_mean = 0;
        this.old_sum = 0;
        this.new_sum = 0;

        return Task.CompletedTask;
    }

    public Task<int> DataValueCount()
    {
        return Task<int>.FromResult(this.count);
    }

    public Task<double> Mean()
    {
        double mean = this.count > 0 ? this.new_mean : 0.0;
        return Task<double>.FromResult(mean);
    }

    public Task<double> Variance()
    {
        double variance = this.count > 1 ? this.new_sum / (this.count - 1) : 0;
        return Task<double>.FromResult(variance);
    }

    public async Task<double> StandardDeviation()
    {
        double variance = await this.Variance();
        return Math.Sqrt(variance);
    }

    public Task Push(double value)
    {
        count++;

        if (count == 1)
        {
            old_mean = value;
            new_mean = value;
            old_sum = 0.0;
        }
        else
        {
            new_mean = old_mean + (value - old_mean) / count;
            new_sum = old_sum + (value - old_mean) * (value - new_mean);
        }

        // set up for next iteration
        old_mean = new_mean;
        old_sum = new_sum;

        return Task.CompletedTask;
    }
}